nexusutils/dataconverter/readers/xps/reader.py

```diff
@@ -0,0 +1,405 @@
+
+            for entry in entry_set:
+                modified_key = key.replace("[entry]", f"[{entry}]")
+                template[modified_key] = field_value
+
+        elif key in list(eln_data_dict.keys()):
+            field_value = eln_data_dict[key]
+            if field_value is None:
+                continue
+            # Do for all entry name
+            for entry in entry_set:
+                modified_key = key.replace("[entry]", f"[{entry}]")
+                # Do for all detector
+                if "[detector]" in key:
+                    for detector in DETECTOR_SET:
+                        detr_key = modified_key.replace("[detector]", f"[{detector}]")
+                        template[detr_key] = field_value
+                else:
+                    template[modified_key] = field_value
+
+
+# pylint: disable=too-few-public-methods
+class XPSReader(BaseReader):
+    """ Reader for XPS.
+    """
+
+    supported_nxdls = ["NXmpes"]
+
+    def read(self,
+             template: dict = None,
+             file_paths: Tuple[str] = None,
+             objects: Tuple[Any] = None) -> dict:
+        """Reads data from given file and returns
+        a filled template dictionary"""
+
+        reader_dir = Path(__file__).parent
+        config_file = reader_dir.joinpath("config_file.json")
+
+        xps_data_dict: Dict[str, Any] = {}
+        eln_data_dict: Dict[str, Any] = {}
+
+        for file in file_paths:
+            file_ext = os.path.splitext(file)[1]
+
+            if file_ext in [".yaml", ".yml"]:
+                with open(file, mode="r", encoding="utf-8") as eln:
+                    eln_data_dict = flatten_and_replace(
+                        FlattenSettings(
+                            yaml.safe_load(eln),
+                            CONVERT_DICT,
+                            REPLACE_NESTED
+                        )
+                    )
+            elif file_ext == ".xml":
+                data_dict = XpsDataFileParser([file]).get_dict()
+                xps_data_dict = {**xps_data_dict, **data_dict}
+
+        with open(config_file, encoding="utf-8", mode="r") as cfile:
+            config_dict = json.load(cfile)
+
+        fill_template_with_xps_data(config_dict,
+                                    xps_data_dict,
+                                    template,
+                                    ENTRY_SET)
+        if eln_data_dict:
+            fill_template_with_eln_data(eln_data_dict,
+                                        config_dict,
+                                        template,
+                                        ENTRY_SET)
+        else:
+            raise ValueError("Eln file must be submited with some required fields and attributes.")
+
+        final_template = Template()
+        for key, val in template.items():
+            if ("/ENTRY[entry]" not in key) and (val is not None):
+                final_template[key] = val
+
+        return final_template
+
+
+READER = XPSReader
```

Comment:
I will vote for having all the functions added for this reader to be part of the class. It fits better and we won't need these ALL_CAPS variables anymore.  @domna what do you think?

Reply:
Yes I agree that we should put the functions in the class. I feel the general pattern for the other readers is different, since we almost always use stand-alone functions but I think it's better to avoid this non-constant uppercase variables (which are also confusing) and have class variables instead.

Reply:
Yes, to avoid confusion, we will have to restructure all the readers.